File: 2022/07/07.rs
aoc2022::main!() ;

fn generator(input : &str) -> Vec<&str>{
    input
        .lines()
        .collect_vec()
}

fn part_1(input : Vec<&str>) -> i32{
    solver(input, 100_000, false)
        .iter()
        .sum()
}

fn part_2(input : Vec<&str>) -> i32 {
    *solver(input.clone(), solver(input.clone(), 0, true).iter().max().unwrap() - 40_000_000, true)
        .iter()
        .min()
        .unwrap()
}

fn solver (input :Vec<&str>, bound : i32, delete : bool) -> Vec<i32> {
    let mut path : Vec<String> = ["root".to_string()].to_vec() ;
    let mut sizes : HashMap<String, i32> = HashMap::new();
    input
        .iter()
        .enumerate()
        .for_each(|(n,command)| {
            let details = command.split(' ').collect_vec() ;
            match details[0] {
            "$" => match details[1..].concat().as_str() {
                "cd.." => path.truncate(&path.len()-1),
                "cd/"  =>  () ,
                "ls"    => () ,
                _       => path.push([details[2] , n.to_string().as_str()].concat().to_string())} 
            "dir"   => (),
            _       =>  path.iter().for_each(|folder|
                            if sizes.contains_key(folder) {
                                sizes.insert(folder.to_string(), details[0].parse::<i32>().unwrap() + sizes[folder]);
                            } else { sizes.insert(folder.to_string(), details[0].parse::<i32>().unwrap()) ; })
                }      
            }
        );
        sizes
        .iter()
        .map(|(folder, _)| sizes[folder])
        .filter(|n| if delete {n > &bound} else {n <= &bound})
        .collect_vec()
        .into()
    
}

File: 2022/07/07.py
def preprocessing(input): 
    return input.splitlines()

def solver(terminal_output):
    small_sizes = 0
    sizes = get_size(terminal_output)
    root_size = sizes.pop()
    for size in sizes:
        if size <= 100_000: small_sizes += size
        if size > (root_size - 40_000_000): 
            break
    yield small_sizes
    yield size
        
def get_size(output):
    sizes = {}
    path =["root"]
    for command in output:
        details = command.split(' ')
        match details: 
            case ["$","cd","/"] : path = ["root"]
            case ["$","cd",".."]: path.pop()
            case ["$","cd",_]   : path.append("".join(path) + details[2])
            case ["dir",_]      : pass 
            case ["$","ls"]     : pass 
            case _: 
                for folder in path: 
                    if folder in sizes: sizes[folder] += int(details[0])
                    else: sizes[folder] = int(details[0])
    return sorted(sizes.values())